test/basic-querying.test.js

```diff
@@ -917,6 +917,71 @@ describe('basic-querying', function() {
           });
         });
     });
+
+    it('applies default values by default', () => {
+      // Backwards compatibility, see
+      // https://github.com/strongloop/loopback-datasource-juggler/issues/1692
+
+      // Initially, all Players were always active, no property was needed
+      const Player = db.define('Player', {name: String});
+      let created;
+      return db.automigrate('Player')
+        .then(() => Player.create({name: 'Pen'}))
+        .then(result => {
+          created = result;
+
+          // Later on, we decide to introduce `active` property
+          Player.defineProperty('active', {
+            type: Boolean,
+            default: false,
```

Comment:
Why not check with a string or a number? The fact that `null` is an acceptable common result for both `applyDefaultsOnReads = true` and `applyDefaultsOnReads  = false` makes the tests a little unreliable in my opinion.

A `. oneOf()` condition might be not be a good choice in this case.

Reply:
> Why not check with a string or a number?

I think it does not really matter. `undefined` or `null` is not a valid value for a boolean in the same way as it is for a string or a number.

> The fact that `null` is an acceptable common result for both `applyDefaultsOnReads = true` and `applyDefaultsOnReads  = false` makes the tests a little unreliable in my opinion.

Please note that SQL databases cannot represent `undefined`, they always use `null`. When `applyDefaultOnReads` is enabled (e.g. by default), then we return the default value, not `null`. Unless you have configured your model property with `default: null`, of course.

Could you please elaborate more on this? I don't understand your concerns :(

> A `.oneOf()` condition might be not be a good choice in this case.

Why not? How do you propose to test handling of default values, considering that different databases are returning different values for properties that have been added to existing model instances (database rows)? SQL databases return `null`, no-SQL databases (memory, MongoDB, etc.) return `undefined`.

Reply:
> Why not? How do you propose to test handling of default values, considering that different databases are returning different values for properties that have been added to existing model instances (database rows)? SQL databases return null, no-SQL databases (memory, MongoDB, etc.) return undefined.

If we don't use a Boolean data type we won't have to deal with the differences in the behavior of different databases. Why bring in a third-party factor in the test case for a subject? If we want to factor in the differences in the behavior of different databases, it should be covered in a separate test case.

Reply:
To sum it up, it looks like to me, that we are testing the behavior of `applyDefaultsOnReads` as well as the database translation of a Boolean value, inside a singe test case. Shouldn't test cases be confined to a single condition, unless it is an acceptance test?

Reply:
I had a chat with @hacksparrow and now I understand the problem - I was using a wrong assertion for the tests checking the legacy behavior where default values are applied. I fixed the problem in f498c26.

Reply:
Actually, it turns out my original test assertions were correct. When the connector returns property value `null`, we don't consider it to be `undefined` and thus the default value is not applied. How confusing! I have updated the comments in the tests to make this more clear.